### ISSUE TITLE
Make keepalive configurable

### ISSUE THREAD

**Issue by lucas-clemente** [lucas-clemente/quic-go#525]
The keepalive proposed in #400 should be configurable via [quic.Config](https://github.com/lucas-clemente/quic-go/blob/master/interface.go#L60). I think we should add a enum there with values Default, Yes, No. Default would be yes for a client and no for a server.

**Comment by krish7919** [comment]
I am interested in contributing to this.

Why not just have a `Keepalive bool` in quic.Config?

From the latest [IETF draft](https://tools.ietf.org/html/draft-hamilton-early-deployment-quic-00):
> 6.10.
  PING frame
   The PING frame can be used by an endpoint to verify that a peer is
   still alive.  The PING frame contains no payload.  The receiver of a
   PING frame simply needs to ACK the packet containing this frame.  The
   PING frame should be used to keep a connection alive when a stream is
   open.  The default is to do this after 15 seconds of quiescence,
   which is much shorter than most NATs time out.  A PING frame only has
   a Frame Type field, and must have the 8-bit Frame Type field set to
   0x07.

1.  Should there also be a configurable timeout required in quic.Config for this? Something like `KeepaliveTimeout int`.

2.  I was going through the source and couldn't find a place where `PingFrame`s are currently being sent. I see that `PingFrame` case is not being handled in the `handleFrames` function in session.go.

Can you confirm that PING Frames are currently not being sent or received?

My idea was to add a call to send ping frames instead of closing session:
```
if now.Sub(s.lastNetworkActivityTime) >= s.idleTimeout() {
    if s.lastNetworkActivityTime > 2*idleTimeout {
       // since we already sent the PING frame and haven't received any activity on the session yet,
       // close the session
        s.close(qerr.Error(qerr.NetworkIdleTimeout, "No recent network activity."))
    } else {
        // since, there is no activity on the session yet, send a PING frame
        <add a call to send PING frame here and not reset lastNetworkActivityTime>
    }   
}
```

**Comment by lucas-clemente** [comment]
Cool, thanks a lot for wanting to help out! :)

We do handle PING frames (https://github.com/lucas-clemente/quic-go/blob/master/packet_unpacker.go#L41, we just don't put them into the slice that is passed on to handlePackets), but we don't currently send them.

The issue with having a simple bool is that we want to be able to use the same config struct for server and client. With bools however we don't have a way of distinguishing between an uninitialized field or the field being set to false. So, that way, we cannot default to different values depending on whether we're a server or client. The rationale here is that only clients should ping, as a server is not  really "interested" in keeping the connection open. Thinking about this some more, an alternative would be to have a bool set to false and switch between keepalive or not only in the http2 mapping. @marten-seeman, thoughts on this?

For the timeout, I generally prefer to have as few knobs as possible. Maybe setting this to half the configured idle state timeout could make sense? We can always add a config option later. What do you think?

Implementation wise, your general idea looks great, but I think a simpler solution should also work: Just sending a PING frame after the keepalive timeout and not touching the idle timeout. Since the PING frame has to be ACKed by the receiver, we should always have some network activity before the idle timeout.

Then, you'd have to add the keepalive timeout to `maybeResetTimer` in the session, and add a check near the end of the main `runLoop`.

Let me know if this clarifies things or if you need more details, and again, thanks! 👍

**Comment by krish7919** [comment]
> The rationale here is that only clients should ping, as a server is not really "interested" in keeping the connection open.

Exactly!

> For the timeout, I generally prefer to have as few knobs as possible. Maybe setting this to half the configured idle state timeout could make sense? We can always add a config option later. What do you think?

You mean a PING frame every `time/2` seconds, and a close in case of no reply at `time` seconds?

> you'd have to add the keepalive timeout to maybeResetTimer in the session, and add a check near the end of the main runLoop.

I didn't understand this, but let me look this up and get back if I have doubts.

**Comment by lucas-clemente** [comment]
> You mean a PING frame every time/2 seconds, and a close in case of no reply at time seconds?

Almost, minor change: a PING frame after time/2 seconds of network inactivity :)

> I didn't understand this, but let me look this up and get back if I have doubts.

Yes, unfortunately this is one of the deepest parts of our implementation. We have a single timer per session, and that timer is set to the next deadline by any part of the protocol (e.g. congestion, retransmission, etc). The timer is one of several things (others being e.g. network activity) that can kick off a run of the session's main run loop, which then has to check which condition caused it to run.

**Comment by lucas-clemente** [comment]
I just discussed this with @marten-seemann on the phone, and we agree that just having a bool instead of an enum in the quic.Config should be okay :) We can then set that to true in the h2 client.

**Comment by krish7919** [comment]
1. Any tips on how to use the `gingko` and `gomega` libraries.

2. If I go with a `Keepalive: true`, should I also add a `KeepaliveTimeout` variable to `quic.Config`. That seems to be a logical inclusion in this case.

**Comment by lucas-clemente** [comment]
re 1: You can find infos on ginkgo [here](https://onsi.github.io/ginkgo/), but don't worry about it too much. This stuff is pretty tricky to test, feel free to just send a PR with a draft and I'll look into it. Roughly (don't worry if this doesn't make sense): I would set the timeout to 0, fire up the session run loop, assert with an `Eventually` that the PING was sent and then shutdown everything.

re 2: Again, I'm not a fan of explicitly specifying the timeout duration. Let me explain this a bit more.

QUIC has the concept of negotiating several transport parameters. One of those is the ICSL, the idle connection state lifetime. If no packets are received for this time, the connection is closed by both peers. (Sidenote: there's talks at the IETF about actually making silent timeouts the default for closing connections.)

The problem is now that for each connection the peers might agree on a different ICSL. So, if the user sets the timeout but not the ICSL, there might be a situation where timeout > ICSL. In that case, keepalive would not work.

Another case for defining the timeout from the ICSL is that we'd like to keep our API minimal (and fields in the Config struct are part of the API). As Go doesn't have dependency versions, if we ever decide to change that field in the future we're essentially breaking all of our clients. So, we're wary of adding stuff unless we absolutely need it.

So, in summary, I'd really prefer to have timeout = ICSL/2. Strictly speaking, you could make the case that it should be more like ICSL/2 - 4*RTT or something, but that just makes things much harder without a benefit for most setups.

**Comment by krish7919** [comment]
Thanks! I will send a rough draft in when I am done and we can finesse it as we go. :)

**Comment by krish7919** [comment]
PR #593.

> you'd have to add the keepalive timeout to maybeResetTimer in the session, and add a check near the end of the main runLoop.

> unfortunately this is one of the deepest parts of our implementation. We have a single timer per session, and that timer is set to the next deadline by any part of the protocol (e.g. congestion, retransmission, etc). The timer is one of several things (others being e.g. network activity) that can kick off a run of the session's main run loop, which then has to check which condition caused it to run.

I still didn't get the above comments.

I do not think I need to reset the timer as a PING frame response does call maybeResetTimer(). Let me know your thoughts if I missed this one.

Also, I have a question regarding the testing: How to run individual tests?

Tests are still failing, and I might not be 100% correct with the gingko tests.
